dclient/cli.py

```diff
@@ -29,17 +30,57 @@ def query(url, sql):
         url = aliases[url]
     if not url.endswith(".json"):
         url += ".json"
-    response = httpx.get(url, params={"sql": sql, "_shape": "objects"})
-    if response.status_code != 200 or not response.json()["ok"]:
+    if token is None:
+        # Maybe there's a token in auth.json?
+        token = _token_for_url_from_auth(url, get_config_dir() / "auth.json")
+    headers = {}
+    if token:
+        headers["Authorization"] = f"Bearer {token}"
+    response = httpx.get(url, params={"sql": sql, "_shape": "objects"}, headers=headers)
+    if response.status_code != 200:
+        # Is it valid JSON?
+        try:
+            data = response.json()
+        except json.JSONDecodeError:
+            raise click.ClickException(
+                "{} status code. Response was not valid JSON".format(
+                    response.status_code
+                )
+            )
+        bits = []
+        if data.get("title"):
+            bits.append(data["title"])
+        if data.get("error"):
+            bits.append(data["error"])
+        raise click.ClickException(
+            "{} status code. {}".format(response.status_code, ": ".join(bits))
+        )
+
+    # We should have JSON now
+    try:
         data = response.json()
+    except json.JSONDecodeError:
+        raise click.ClickException("Response was not valid JSON")
+    # ... but it may have a {"ok": false} error
+    if not data.get("ok"):
         bits = []
         if data.get("title"):
             bits.append(data["title"])
         if data.get("error"):
             bits.append(data["error"])
+        if not bits:
+            bits = [json.dumps(data)]
         raise click.ClickException(": ".join(bits))
-    else:
-        click.echo(json.dumps(response.json()["rows"], indent=2))
+
+    # Output results
+    click.echo(json.dumps(response.json()["rows"], indent=2))
+
+
+def _token_for_url_from_auth(url, auth_file):
+    auths = _load_auths(auth_file)
+    for auth_url, token in auths.items():
+        if url.startswith(auth_url):
+            return token
```

Comment:
This feels delicate. I'd like this to at least use proper URL parsing to match on hostname and path rather than a dumb `.startswith()` check.

Reply:
Fixed.